enterprise/server/build_event_publisher/build_event_publisher.go

```diff
@@ -88,98 +93,133 @@ func (p *Publisher) run(ctx context.Context) {
+		p.events.Wait()
+	}
+
+	// Return any error from receiving ACKs
+	return <-doneReceiving
+}
 
+func (p *Publisher) Publish(event *bespb.BuildEvent) error {
+	be := &bepb.BuildEvent{EventTime: ptypes.TimestampNow()}
+	if event == nil {
+		be.Event = &bepb.BuildEvent_ComponentStreamFinished{ComponentStreamFinished: &bepb.BuildEvent_BuildComponentStreamFinished{
+			Type: bepb.BuildEvent_BuildComponentStreamFinished_FINISHED,
+		}}
+	} else {
 		bazelEvent, err := ptypes.MarshalAny(event)
 		if err != nil {
-			p.setError(status.WrapError(err, "failed to marshal bazel event"))
-			return
-		}
-		start := time.Now()
-		err = stream.Send(&pepb.PublishBuildToolEventStreamRequest{
-			OrderedBuildEvent: &pepb.OrderedBuildEvent{
-				StreamId:       p.streamID,
-				SequenceNumber: seqNo,
-				Event: &bepb.BuildEvent{
-					EventTime: ptypes.TimestampNow(),
-					Event:     &bepb.BuildEvent_BazelEvent{BazelEvent: bazelEvent},
-				},
-			},
-		})
-		log.Debugf("BEP: published event (#%d) in %s", seqNo, time.Since(start))
-		if err != nil {
-			p.setError(err)
-			return
+			return err
 		}
+		be.Event = &bepb.BuildEvent_BazelEvent{BazelEvent: bazelEvent}
 	}
-	// After successfully transmitting all events, close our side of the stream
-	// and wait for server ACKs before closing the connection.
-	if err := stream.CloseSend(); err != nil {
-		p.setError(status.WrapError(err, "failed to close build event stream"))
-		return
-	}
-	<-doneReceiving
+
+	p.events.Add(be)
+	return nil
 }
 
-func (p *Publisher) Publish(e *bespb.BuildEvent) error {
-	p.mu.Lock()
-	defer p.mu.Unlock()
-	if p.err != nil {
-		return status.WrapError(p.err, "cannot publish event due to previous error")
+// Wait publishes the Finished event and then waits for the server to ACK
+// all published events.
+func (p *Publisher) Wait() error {
+	if err := p.Publish(nil); err != nil {
+		return err
+	}
+	if err := <-p.done; err != nil {
+		return status.UnavailableErrorf("failed to publish build event stream after multiple attempts: %s", err)
 	}
-	p.events <- e
 	return nil
 }
 
-func (p *Publisher) Wait() error {
-	p.events <- nil
-	<-p.done
-	return p.err
+// EventBuffer holds build events that have not been ack'd by the server.
+//
+// It supports a single consumer and multiple producers: multiple goroutines
+// can call Add(), and a single goroutine can call Get() and Wait() in a loop
+// to observe added events. The consumer keeps track of which events it has
+// already processed.
+type EventBuffer struct {
+	mu sync.Mutex
+	// TODO(bduffany): Store these on disk since the buffer can get quite large if
+	// there is a lot of log output.
+	items []*bepb.BuildEvent
+	// notify is used to signal that an event has been added to the buffer.
+	notify chan struct{}
+}
+
+func NewEventBuffer() *EventBuffer {
+	return &EventBuffer{notify: make(chan struct{}, 1)}
+}
+
+// Get returns the currently buffered items.
+func (b *EventBuffer) Get() []*bepb.BuildEvent {
```

Comment:
It would be a little cleaner to have a `Subscribe()`-like method that returns a new channel that streams the data from `items` (existing and new) instead of putting that responsibility on the client.

Reply:
Done -- I agree that this is cleaner :)